app/forms/otp_verification_form.rb

```diff
@@ -13,6 +13,8 @@ def submit
   attr_reader :code, :user
 
   def valid_direct_otp_code?
+    code_length = Devise.direct_otp_length
+    return false unless code =~ /^\d{#{code_length}}$/
```

Comment:
For defensive code like this, how about anchoring with `\A` and `\Z` as beginning and end of input (rather than line)?

This protects against nefarious inputs with newlines:

```ruby
"aaaaa\n123456\naaaaaaaaa" =~ /^\d{6}$/
=> 6
"aaaaa\n123456\naaaaaaaaa" =~ /\A\d{6}\Z/
=> nil
```